File: leetcode/Convert_Sorted_Array_to_Binary_Search_Tree.java
/**
 * Definition for binary tree
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    int[] num;
    
    TreeNode sortedArrayToBST(int a, int b) {
        if (a > b) return null;
        int m = a + (b - a + 1) / 2;
        TreeNode tn = new TreeNode(num[m]);
        tn.left = sortedArrayToBST(a, m - 1);
        tn.right = sortedArrayToBST(m + 1, b);
        return tn;
    }
    
    public TreeNode sortedArrayToBST(int[] __num) {
        num = __num;
        return sortedArrayToBST(0, num.length - 1);
    }
}

File: leetcode/Convert_Sorted_Array_to_Binary_Search_Tree.cpp
/**
 * Definition for binary tree
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
    TreeNode *buildBST(vector<int> &num, int begin, int end) {
        if (begin == end)
            return NULL;
        int mid = (begin + end) / 2;
        TreeNode *tree_node = new TreeNode(num[mid]);
        tree_node->left = buildBST(num, begin, mid);
        tree_node->right = buildBST(num, mid + 1, end);
        return tree_node;
    }
public:
    TreeNode *sortedArrayToBST(vector<int> &num) {
        return buildBST(num, 0, num.size());
    }
};
